src/server/bg_services/lifecycle.js

```diff
@@ -11,63 +11,113 @@ const system_store = require('../system_services/system_store').get_instance();
 const auth_server = require('../common_services/auth_server');
 
 const LIFECYCLE = {
-    schedule_min: 1 //run every 5 minutes
+    schedule_min: 5 * 1000 * 60 // run every 5 minutes
 };
 
+function get_expiration_timestamp(expiration) {
+    if (!expiration) {
+        return undefined; // undefined
+    } else if (expiration.date) {
+        return Math.floor(new Date(expiration.date).getTime() / 1000);
+    } else if (expiration.days) {
+        return moment().subtract(expiration.days, 'days').unix();
+    }
+}
+
+/*
+*
+* S3 ExpiredObjectDeleteMarker logic
+*        If set to true, the delete marker will be expired; if set to false the policy takes no action.
+*
+* MDStore find_objects() filter_delete_markers logic:
+*       if (filter_delete_markers === true) delete_marker = null;
+*       else if (filter_delete_markers === false) delete_marker = true;
+*/
+function get_expiration_filter_delete_markers(expiration) {
+    if (expiration && expiration.expired_object_delete_marker) {
+        return false; // delete_marker = true
+    }
+    return true; // delete_marker = null
+}
+
+function get_filter_field(filter, field) {
+    return filter ? filter[field] : undefined;
+}
+
+function get_prefix(filter) {
+    return get_filter_field(filter, 'prefix');
+}
+
+function get_size_less_than(filter) {
+    return get_filter_field(filter, 'object_size_less_than');
+}
+
+function get_size_greater_than(filter) {
+    return get_filter_field(filter, 'object_size_greater_than');
+}
+
+function get_tags(filter) {
+    return get_filter_field(filter, 'tags');
+}
+
+async function handle_bucket_rule(system, rule, j, bucket) {
+    dbg.log0('LIFECYCLE HANDLING BUCKET:', bucket.name, '(bucket id:', bucket._id, ') BEGIN');
+    const now = Date.now();
+    const bucket_rule = bucket.name + '(bucket id:' + bucket._id + ') rule id(' + j + ') ' + rule.id;
+
+    if (rule.status !== 'Enabled') {
+        dbg.log0('LIFECYCLE SKIP bucket:', bucket.name, '(bucket id:', bucket._id, ') rule id', rule.id, 'rule', rule, 'not Enabled');
+        return;
+    }
+    if (rule.last_sync && now - rule.last_sync < LIFECYCLE.schedule_min) {
+        dbg.log0('LIFECYCLE SKIP bucket:', bucket.name, '(bucket id:', bucket._id, ') rule id', rule.id, 'rule', rule, 'now', now, 'last_sync', rule.last_sync, 'schedule min', LIFECYCLE.schedule_min);
+        return;
+    }
+    if (rule.expiration === undefined) {
+        dbg.log0('LIFECYCLE SKIP bucket:', bucket.name, '(bucket id:', bucket._id, ') rule id', rule.id, 'rule', rule, 'now', now, 'last_sync', rule.last_sync, 'no expiration');
+        return;
+    }
+    dbg.log0('LIFECYCLE PROCESSING bucket:', bucket_rule);
+
+    await server_rpc.client.object.delete_multiple_objects_by_filter({
+        bucket: bucket.name,
+        create_time: get_expiration_timestamp(rule.expiration),
+        filter_delete_markers: get_expiration_filter_delete_markers(rule.expiration),
+        prefix: get_prefix(rule.filter),
```

Comment:
the getter seems unnecessary here. if it's just to check if `filter` is defined, you can either use optional chaining - `rule?.filter?.prefix` (this syntax should be supported in node 14) or just use `rule.filter && rule.filter.prefix`.
The same goes for the other getters

```suggestion
        prefix: rule?.filter?.prefix,
```
or 
```suggestion
        prefix: rule.filter && rule.filter.prefix,
```

Reply:
👍